### ISSUE TITLE
Exception hierarchy under FeignException

### ISSUE THREAD

**Issue by jerzykrlk** [OpenFeign/feign#823]
With default error encoder, Feign client will throw `FeignException`, which is quite general, compared to variety of HTTP error codes and their REST semantics.
In a rest-heavy application, introducing subclasses simplify quite a lot of my code. 

Sample piece of code, now:

```java
        try {
            return libraryClient.getBook(1);
        } catch (FeignException e) {
            if (e.status() == 404) {
                // handle the 404 error
            }
            throw e; // can't do anything about other http errors here
        }
```

And the same piece code, expected:

```java
        try {
            return libraryClient.getBook(1);
        } catch (HttpNotFoundException e) {
            // handle the 404 error
        }
```

This change seems quite easy, compared ti the code it saves. I could prepare a pull request, if you like the idea.

**Comment by navaare** [comment]
@jerzykrlk 
How much code does it save in fact? Does everybody will be happy with that default behavior? All you need is to define your own exception passing a lot more info at defined ErrorDecoder for feign client.

**Comment by jerzykrlk** [comment]
Hi,

I think the main benefit would not be less code, but just clener code. 

In my apps, there are many places where I catch `FeignException`, manually check the status code - and if it's something else, rethrow the exception. 

For example: 
- a call to a remote service returns HTTP 401,
- my code catches `FeignException`, checks for status 401, reauthenticates,
- otherwise - rethrows the `FeignException`.

I think it doesn't look good, especially with the same thing repeated in many places in code in many apps.

I already have a custom `ErrorDecoder` that does that - but I believe it'd be a good add-on to the `Default` error decoder.

This whole idea is a follow-up of the same feature in Spring `RestTemplate` - see [SPR-15404](https://jira.spring.io/browse/SPR-15404) and [this commit](https://github.com/spring-projects/spring-framework/commit/7f0e3481a562a8bf63efced9ef2b06712ac573bf)

All HTTP-related exceptions would inherit from `FeignException`, so the change would be backward compatible.

Could you have a look at https://github.com/OpenFeign/feign/pull/825? - I think it illustrates my idea clearly.

**Comment by kdavisk6** [comment]
The role of the `ErrorDecoder` is to allow users to maintain full control.  I do see the benefit, but it may not be Feign's place to make this determination across the board for all users nor desired.  In practice, we typically do not accept enhancement requests without additional votes.  I've marked this issue as such.  If it receives enough support, we can consider this more.

**Comment by jerzykrlk** [comment]
Ok, thank you for the update. 

You already have first vote, I guess... :)

**Comment by kdavisk6** [comment]
Closed via #825
